### ISSUE TITLE
Multicore evaluation doesn't work on Windows OS

### ISSUE THREAD

**Issue by mikefenton** [PonyGE/PonyGE2#45]
For some reason, multicore evaluation on Windows machines reverts the params dictionary back to the default state (i.e. the default params dictionary specified in src/algorithm/parameters.py) within src/representation/individiaul/Individiaul.evaluate. This means that when Individual.evaluate tries to execute params['FITNESS_FUNCTION'] it is trying to execute a string rather than a class object. This only happens on Windows machines when --multicore is used.

**Comment by mikefenton** [comment]
To provide an update on this, the crux of the issue is with Python's forking process in the `Multiprocessing` module. Essentially, Windows platforms don't have a `fork()` system call. On Unix systems, child processes fork off the main parent process and retain the system memory of the parent process at the time of their creation. However, the default Multiprocessing operation on Windows machines is `spawn()`. This means that when a child process is spawned, any modules it uses that aren't explicitly passed in at the time of creation by the parent process are re-created from scratch by the child process. Essentially, Windows `Multiprocessing` instances re-set all system memory to that which is explicitly defined _at the time of initialisation_.

This becomes an issue for us in PonyGE2, as we store all of our statistics, parameters, and population trackers centrally rather than passing them explicitly in to various functions. On Windows machines, any calls to these central repositories are making calls back to their original initialised state (i.e., the params dictionary as specified directly in the code in `algorithm.parameters.params`). Since we pass in parameters from parameters files or from the command line, these specified values are essentially "ignored" by the spawned child process, as the child process imports a fresh new copy of the params dictionary (and any other modules imported or used in the `Individual.individual.evaluate()` call) rather than the copy that exists in memory.

Another issue that arises because of this is that module objects are no longer callable within fitness functions. This means that class instance variables (e.g. `self.whatever`) that are defined during initialisation are not defined in fitness function `evaluate()` methods, and raise further errors. Obviously this raises all sorts of further problems, such as not being able to load in datasets during initialisation. This also precludes the possibility of us passing in all required modules to the `Individual.individual.evaluate()` call (e.g. the current state of the `algorithm.parameters.params` and `stats.stats.stats` dictionaries, and all info contained in `utilities.stats.trackers`). Basically, since Windows doesn't have the necessary `fork()` system call, I'm not sure what we can do about this.

See here: https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods

**Comment by t-h-e** [comment]
Multicore functionality for Windows has been added with commit: https://github.com/PonyGE/PonyGE2/commit/e4ec5d18186e8da714ef10191b4416a3d753ee29

This commit also changes the way the `Pool` for multicore evaluation is created. Instead of creating a new `Pool` every generation, the `Pool` is created only one time within the search loop.

A problem that still exists, are race conditions. One that I found e.g.: https://github.com/PonyGE/PonyGE2/blob/master/src/fitness/base_ff_classes/moo_ff.py#L63
There may be more.
